Code/RDGeneral/ConcurrentQueue.h

```diff
@@ -0,0 +1,132 @@
+//
+//  Copyright (C) 2020 Shrey Aryan
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#ifdef RDK_THREADSAFE_SSS
+#ifndef CONCURRENT_QUEUE
+#define CONCURRENT_QUEUE
+#include <condition_variable>
+#include <thread>
+#include <vector>
+
+namespace RDKit {
+template <typename E>
+class ConcurrentQueue {
+ private:
+  unsigned int d_capacity;
+  bool d_done;
+  std::vector<E> d_elements;
+  unsigned int d_head, d_tail;
+  mutable std::mutex d_lock;
+  std::condition_variable d_notEmpty, d_notFull;
+
+ private:
+  ConcurrentQueue<E>(const ConcurrentQueue<E>&);
+  ConcurrentQueue<E>& operator=(const ConcurrentQueue<E>&);
+
+ public:
+  ConcurrentQueue<E>(unsigned int capacity)
+      : d_capacity(capacity), d_done(false), d_head(0), d_tail(0) {
+    std::vector<E> elements(capacity);
+    d_elements = elements;
+  }
+
+  //! tries to push an element into the queue if it is not full without
+  //! modifying the variable element, if the queue is full then pushing an
+  //! element will result in blocking
+  void push(const E& element);
+
+  //! tries to pop an element from the queue if it is not empty and not done
+  //! the boolean value indicates the whether popping is successful, if the
+  //! queue is empty and not done then popping an element will result in
+  //! blocking
+  bool pop(E& element);
+
+  //! checks whether the ConcurrentQueue is empty
+  bool isEmpty() const;
+
+  //! returns the value of the variable done
+  bool getDone() const;
+
+  //! sets the variable d_done = true
+  void setDone();
+
+  //! clears the vector
+  void clear();
+};
+
+template <typename E>
+void ConcurrentQueue<E>::push(const E& element) {
+  std::unique_lock<std::mutex> lk(d_lock);
+  //! concurrent queue is full so we wait until
+  //! it is not full
+  while (d_head + d_capacity == d_tail) {
+    d_notFull.wait(lk);
```

Comment:
You have a dead lock here.  If we push an item, we lock the queue and THEN wait on for the queue to be emptied, which can't happen since we hold the lock.

A small change to testProducerConsumer shows this where we call thread::join before launching the consumers.

```
  //! start producer threads
  for (int i = 0; i < numProducerThreads; i++) {
    producers[i] = std::thread(produce, std::ref(q), numToProduce);
  }

  std::for_each(producers.begin(), producers.end(),
                std::mem_fn(&std::thread::join));


  //! start consumer threads
  for (int i = 0; i < numConsumerThreads; i++) {
    consumers[i] = std::thread(consume, std::ref(q), std::ref(results[i]));
  }

  //! the producer is done producing
  q.setDone();

  std::for_each(consumers.begin(), consumers.end(),
                std::mem_fn(&std::thread::join));
  TEST_ASSERT(q.isEmpty());
```

Reply:
@bp-kelley, assuming I understand your comment correctly:
The use of the condition_variable `d_notFull` ensures that we don't get a deadlock here. We no longer hold the lock while `d_notFull.wait()` is executing.

Reply:
The above test was wrong as the join was blocking.  I added a sleep in between the producers and consumers and it looks ok now, although I did fail the assert once (q.isempty) I'm assuming that was a mistake on my part for now.